Virtual Observatory registry query (#1158)
astropy/astroquery: docs/vo_service_discovery/vo_service_discovery.rst

provided by the NASA Extragalactic Database (NED).

.. code-block:: python

    >>> from astroquery.vo_service_discovery import Registry
    >>> results = Registry.query(source='ned', service_type='cone')
    >>> print(results)

                                waveband                         ... service_type
        -------------------------------------------------------- ... ------------
                                                        infrared ...   conesearch
        radio#millimeter#infrared#optical#uv#euv#x-ray#gamma-ray ...   conesearch

The result of the query is an `astropy.table.Table` object. The methods of `Table`
can be used to inspect the search result.

.. code-block:: python
    >>> print(results.colnames)

        ['waveband', 'short_name', 'ivoid', 'res_description', 'access_url',
         'reference_url', 'publisher', 'service_type']

The value of `'access_url'` specifies the endpoint for the service.

Documentation of the service may be found in the `'reference_url'` values.
We could use capabilities in the standard library to open the first url
in a web browser.

.. code-block:: python

    >>> import webbrowser

>     >>> webbrowser.open(results['reference_url'][0])

Reviewer: In the course of testing, I modified this to:
```suggestion
    >>> webbrowser.open(results['reference_url'][1].decode())
```
because the item is a byte-string, which  causes an exception in `web browser.open`.

ways to search for specific services of interest. Each of these options
can be combined into a single call to the `query` method.

The ``service_type`` parameter can be set to the VO service types ``'conesearch'``,
``'simpleimageaccess'``, ``'simplespectralaccess'``, or ``'tableaccess'``. These
values may be shortened to ``'cone'``, ``'image'``, ``'spectra'`` or
``'spectrum'``, or ``'tap'`` or ``'table'``.

The value of the ``keyword`` parameter will return results containg that string
in the ``ivoid``, ``short_name``, or ``res_description`` fields.

Wavelength range can be specified by a string containing one or more band names.
The options include gamma-ray (wavelength below 0.1 Angstrom), x-ray
(0.1 to 100 Angstrom), euv (100-1000 Angstrom), uv (1000 to 3000 Angstrom),
optical (3000 to 10000 Angstrom), infrared (1 to 100 micron), millimeter (0.1 to
10 mm), or radio (greater than 10 mm wavelength). Multiple bands can be specified
with a comma-separated list.

The ``source`` parameter is a string that is matched to any substring in ivoid.

The ``publisher`` is the name of any publishing organization.

``order_by`` can be specified to sort the result by any of the column names.

> The ``logic_string`` is any other string that is needed for the ADQL ``where``

Reviewer: Delete this paragraph. `logic_string` is not in the function signature.

clause, and must start with the value ``' and '``.

The ADQL query can be printed by specifying ``verbose=True``.

> 

Reviewer: Add to this the other options of `query`.

```suggestion

The results of the query can be saved to a file. Set ``dump_to_file=True`` to save the results table. The format will be ``'votable'`` by default; you can change this with the ``output_format`` option. The result will be saved to a filename using the TAP job ID; you can specify the filename with the ``output_file`` option.

By default, the query is made using the  registry TAP URL. The ``url`` option will override that value. Alternatively, the configuration can be changed to use a different URL.
```

Author: Addressed.

.. code-block:: python

    >>> result = Registry.query(service_type='tap', source='heasarc', verbose=True)

    Registry:  sending query ADQL =
              select res.waveband,res.short_name,cap.ivoid,res.res_description,
              intf.access_url,res.reference_url,res_role.role_name as publisher,cap.cap_type as service_type
              from rr.capability as cap
                natural join rr.resource as res
                natural join rr.interface as intf
                natural join rr.res_role as res_role
                 where cap.cap_type like '%tableaccess%' and cap.ivoid like '%heasarc%' and res_role.base_role = 'publisher'

    Queried: https://vao.stsci.edu/RegTAP/TapService.aspx/sync

    >>> print(result)

    waveband short_name ...  publisher   service_type
    -------- ---------- ... ------------ ------------
                HEASARC ... NASA/HEASARC  tableaccess

Counting available services
===========================

The `query_counts` method will enumerate how many services are available

> for each of the fields in the results table shown above.

Reviewer: ```suggestion
for the ``'service_type'``, ``'waveband'`` or ``'publisher'`` fields.
```

Author: Addressed.

.. code-block:: python

    >>> result = Registry.query_counts(field='service_type')
    >>> print(result)

           service_type       count_service_type
    ------------------------- ------------------
                   capability              59214
                   conesearch              21453
            simpleimageaccess                334
         simplespectralaccess                115
                  tableaccess                 97
                      edition                 86
                ceacapability                 56
                      harvest                 41
             simpletimeaccess                 31
                  openskynode                 27
                       search                  8
    theoreticalspectralaccess                  5
                     heliotap                  5
          protospectralaccess                  4
             simplelineaccess                  3

> 

Reviewer: Document the options.

```suggestion

The ``minimum`` option defaults to 1 and specifies the minimum counts needed for a row to be listed in the results table. `query_counts` includes the same ``verbose``, ``url`` and file output options as the ``query`` method.

```

Author: Addressed.

> We can get a list of the top 15 publishers:

Reviewer: The code block is for all publishers with at least 15 services, not the top 15 publishers.

```suggestion
We can get a list of all publishers with at least 15 services.
```

Author: Addressed.

.. code-block:: python

>     >>> results = Registry.query_counts('publisher', 15, verbose=True)

Reviewer: Recommend to specify the option `minimum`.

```suggestion
    >>> results = Registry.query_counts('publisher', minimum=15, verbose=True)
```

Author: Addressed.

.. doctest-skip-all

.. _astroquery.vo_service_discovery:

*************************************************************************
Virtual Observatory Service Discovery (`astroquery.vo_service_discovery`)
*************************************************************************

Getting Started
===============

Archives from around the world have registered their services with the NASA
Astronomy Virtual Observatory (NAVO) directory. A web interface is available
at http://vao.stsci.edu/keyword-search/. This module provides methods
to query this directory by service type, keywords, wavelength band or coverage,
and source or publisher.

Here is an example showing how to find information about conesearch services
provided by the NASA Extragalactic Database (NED).

.. code-block:: python

>     >>> from astroquery.vo_service_discovery import Registry

Reviewer 2: One big-picture suggestion: can we either rename the module `VORegistry` or suggest a user standard of doing that?   e.g., `from astroquery.vo_service_discovery import Registry as VORegistry`?  I don't like the idea of user code existing in the wild with the generic name `Registry`, since that could be a registry for anything.

    >>> results = Registry.query(source='ned', service_type='cone')
    >>> print(results)

                                waveband                         ... service_type
        -------------------------------------------------------- ... ------------
                                                        infrared ...   conesearch
        radio#millimeter#infrared#optical#uv#euv#x-ray#gamma-ray ...   conesearch

The result of the query is an `astropy.table.Table` object. The methods of
`~astropy.table.Table`
can be used to inspect the search result.

.. code-block:: python

    >>> print(results.colnames)

        ['waveband', 'short_name', 'ivoid', 'res_description', 'access_url',
         'reference_url', 'publisher', 'service_type']

The value of ``'access_url'`` specifies the endpoint for the service.

Documentation of the service may be found in the ``'reference_url'`` values.
You can use a function from the Python standard library to open the second url
in a web browser.

.. code-block:: python

    >>> import webbrowser

>     >>> webbrowser.open(results['reference_url'][1].decode())

Reviewer 3: This is basically a safety measure as not all URLs across all registries are stored decoded. While it would be best if the user did not have to decode any results, there is little explicit documentation within the IVOA standards about whether all the URLs should be escaped or not, and some inconsistency remains among the ~20k current resources.

>     >>> webbrowser.open(results['reference_url'][1].decode())

Reviewer 4: IMHO such technicalities should be hidden from users if possible. But if it is impossible, perhaps an explanation in the doc on why decode is needed here?

>     >>> webbrowser.open(results['reference_url'][1].decode())

Author: The commit comment for the decode() says it's needed because the reference_url value is a byte-string.  I think that's an issue with the VOTable parser in Astropy, but a little more looking is needed.  

If the reference_url were a str, it may not need decoding regardless of whether the value were already URL-encoded.

https://github.com/astropy/astroquery/pull/1158#discussion_r240796583

>     >>> webbrowser.open(results['reference_url'][1].decode())

Reviewer 3: Apologies, I should have looked closer. There have been other places where we safety-decode URLs in strings from the registry and hide it from the user, I think Tom's right that this is a separate thing.

>     >>> webbrowser.open(results['reference_url'][1].decode())

Reviewer 4: I won't be surprised if this stemmed from `io.votable`, as that package needs some maintenance. 😅

>     >>> webbrowser.open(results['reference_url'][1].decode())

Reviewer 2: Where did we end up coming down on this?  I really don't like the idea of telling users to use `decode`.  I'd slightly prefer a hack-fix to the votable parser within this module over documentation telling users to change string types.

Discovering Services
====================

The `~astroquery.vo_service_discovery.RegistryClass.query` method provides several
ways to search for specific services of interest. Each of these options
can be combined into a single call to the method.

The ``service_type`` parameter can be set to the VO service types ``'conesearch'``,
``'simpleimageaccess'``, ``'simplespectralaccess'``, or ``'tableaccess'``. These
values may be shortened to ``'cone'``, ``'image'``, ``'spectra'`` or
``'spectrum'``, or ``'tap'`` or ``'table'``.

> The value of the ``keyword`` parameter will return results containg that string

Reviewer 2: ```suggestion
The value of the ``keyword`` parameter will return results containing that string
```

in the ``ivoid``, ``short_name``, or ``res_description`` fields.

> Wavelength range can be specified by a string containing one or more band names.

Reviewer 2: this paragraph doesn't define the keyword needed here; presumably it's `wavelength`?

Reviewer 2: no, it's `waveband`.

Reviewer 2: also, maybe combine the last sentence with the first?

Author: I reworded the waveband section to try to address the comments.

Regarding the question, VO registries do not currently have a working ability to query for services that have overlap with some part of the sky.  

However, the IVOA is working on enhancements to allow services to describe their sky coverage and to support queries based on that.  Those features would likely be based on Multi-Order Coverage maps, much like the CDS MOC Server does today (https://astroquery.readthedocs.io/en/latest/api/astroquery.cds.CdsClass.html#astroquery.cds.CdsClass.query_region)
